pkg/helm/helm_helpers.go

```diff
@@ -0,0 +1,89 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helm
+
+import (
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kanisterio/kanister/pkg/field"
+	"github.com/kanisterio/kanister/pkg/log"
+)
+
+type k8sObj struct {
+	ObjKind  string            `json:"kind"`
+	MetaData metav1.ObjectMeta `json:"metadata"`
+}
+
+type K8sObjectType string
+
+type RenderedResource struct {
+	name             string
+	renderedManifest string // This holds the dry run string output of the resource.
+}
+
+type ResourceFilter func(kind K8sObjectType) bool
+
+// ResourcesFromRenderedManifest extracts and optionally filters rendered resources from a given rendered manifest.
+// Returns:
+// - A slice of RenderedResource structs containing the name and the YAML of each resource that passes the filter.
+func ResourcesFromRenderedManifest(manifest string, filter ResourceFilter) []RenderedResource {
+	var ret []RenderedResource
+	// Get rid of the notes section, shown at the very end of the output.
+	manifestSections := strings.Split(manifest, "NOTES:")
+	// The actual rendered manifests start after first occurrence of `---`.
+	// Before this we have chart details, for example Name, Last Deployed, Status etc.
+	renderedResourcesYaml := strings.Split(manifestSections[0], "---")
+	for _, resourceYaml := range renderedResourcesYaml[1:] {
+		obj := k8sObj{}
+		if err := yaml.Unmarshal([]byte(resourceYaml), &obj); err != nil {
+			log.Error().Print("Failed to unmarshal k8s obj", field.M{"Error": err})
+			continue
+		}
+		k8sType := K8sObjectType(strings.ToLower(obj.ObjKind))
+		// Either append all rendered resource or filter.
+		if filter == nil || filter(k8sType) {
+			ret = append(ret, RenderedResource{
+				name:             obj.MetaData.Name,
+				renderedManifest: resourceYaml,
+			})
+		}
+	}
+	return ret
+}
+
+// GetK8sObjectsFromRenderedManifest unmarshals a list of rendered Kubernetes manifests
+// into a map of Kubernetes objects. It unmarshals each manifest into an object of the specified type `T`,
+// which must implement the runtime.Object interface.
+//
+// Returns:
+//   - A map where the keys are the names of the resources and the values are the unmarshaled K8S objects of type `T`.
+//   - An error if any manifest fails to unmarshal.
+func GetK8sObjectsFromRenderedManifest[T runtime.Object](resources []RenderedResource) (map[string]T, error) {
```

Comment:
```suggestion
func K8sObjectsFromRenderedResources[T runtime.Object](resources []RenderedResource) (map[string]T, error) {
```

we are getting k8s objects from rendered resources and not the rendered manifest right?

Reply:
Done